src/backend/optimizer/prep/prepunion.c

```diff
@@ -1808,6 +1808,54 @@ adjust_appendrel_attrs(PlannerInfo *root, Node *node, AppendRelInfo *appinfo)
 	return result;
 }
 
+typedef struct nested_subplans_context
+{
+	plan_tree_base_prefix base;
+	PlannerInfo *root;
+} nested_subplans_context;
+
+static bool
+nested_subplans(Node *node, adjust_appendrel_attrs_context *context) {
+	nested_subplans_context *new_context = (nested_subplans_context*)malloc(sizeof(nested_subplans_context));
+	new_context->root = context->root;
```

Comment:
Maybe move this logic to `adjust_appendrel_attrs_mutator` function?

Reply:
I assume that the usage of wraparounds is common in gpdb, so the intention of nested_subplans is to hide the conversion from adjust_appendrel_attrs_mutator and keep it tidy. 
Regerding malloc, can you suggest a common allocator, because as I can see, all the allocators are type-specific?

Reply:
I don't think that this walker could be used somewhere else.
Please read article about Postgres memory allocation: src/backend/utils/mmgr/README
In this case, we can put struct item on stack and avoid memory allocation at all.

Reply:
I swithced from malloc to palloc and used pfree after `nested_subplan_mutator`. I also removed a wraparound, and the allocation is now located in `adjust_appendrel_attrs_mutator`

Reply:
Do we really need memory allocation?

> In this case, we can put struct item on stack and avoid memory allocation at all.

Why do you leave `base->node` undefined?

Reply:
> Why do you leave `base->node` undefined?
The `node` field can have different types and the name of this field doesn't reflect the type of its content. I added the `root` field to `nested_subplans_context` to keep it similar to `adjust_appendrel_attrs_mutator`.

Reply:
> Structure plan_tree_base_prefix carries the appropriate pointer for GPDB's general plan tree walker/mutator framework.  **All users** of the framework **must prefix** their context structure with a plan_tree_base_prefix and **initialize** it appropriately.

according to the structure [comment](https://github.com/arenadata/gpdb/blob/619755cdcd8c197e9fb8339a5b33ad80e02f9995/src/include/optimizer/walkers.h#L16-L27). Of course, currently, you handle SubPlan nodes separately in your function. But agreements were invented to simplify code maintenance.